pkg/virtctl/credentials/credentials.go

```diff
@@ -0,0 +1,239 @@
+			if err != nil {
+				return fmt.Errorf("error patching secret \"%s\": %w", secretName, err)
+			}
+
+			cmd.Printf("Successfully added the key to secret \"%s\"", secretName)
+			return nil
+		},
+	}
+	cmdFlags.AddToCommand(cmd)
+
+	return cmd
+}
+
+type sshCommandFlags struct {
+	user             string
+	secret           string
+	sshPubKeyFile    string
+	sshPubKeyLiteral string
+}
+
+const (
+	userFlag     = "user"
+	keyFileFlag  = "file"
+	keyValueFlag = "value"
+)
+
+func (s *sshCommandFlags) AddToCommand(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&s.user, userFlag, "u", "", "Name of the user.")
+	err := cmd.MarkFlagRequired(userFlag)
+	if err != nil {
+		panic(err)
+	}
+
+	cmd.Flags().StringVar(&s.secret, "secret", "", "Name of the secret with SSH keys.")
+
+	cmd.Flags().StringVarP(&s.sshPubKeyFile, keyFileFlag, "f", "", "Path to the SSH public key file.")
+	err = cmd.MarkFlagFilename(keyFileFlag)
+	if err != nil {
+		panic(err)
+	}
+
+	cmd.Flags().StringVar(&s.sshPubKeyLiteral, keyValueFlag, "", "Literal value of the SSH public key.")
+	cmd.MarkFlagsMutuallyExclusive(keyFileFlag, keyValueFlag)
+}
+
+func getSshKey(flags *sshCommandFlags) (string, error) {
+	if flags.sshPubKeyLiteral != "" {
+		return flags.sshPubKeyLiteral, nil
+	}
+
+	if flags.sshPubKeyFile == "" {
+		return "", fmt.Errorf("one of --%s, or --%s must be specified", keyFileFlag, keyValueFlag)
+	}
+	data, err := os.ReadFile(flags.sshPubKeyFile)
+	if err != nil {
+		return "", err
+	}
+	return string(data), nil
+}
+
+func getAccessCredentialsForVm(ctx context.Context, cli kubecli.KubevirtClient, namespace string, name string) ([]v1.AccessCredential, error) {
+	vmi, err := cli.VirtualMachineInstance(namespace).Get(ctx, name, &metav1.GetOptions{})
+	if err == nil {
+		return vmi.Spec.AccessCredentials, nil
+	}
+	if !errors.IsNotFound(err) {
+		return nil, err
+	}
+
+	vm, err := cli.VirtualMachine(namespace).Get(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	return vm.Spec.Template.Spec.AccessCredentials, nil
+}
+
+func getSshSecretsForUser(accessCredentials []v1.AccessCredential, user string) []string {
+	var result []string
+	for i := range accessCredentials {
+		credential := &accessCredentials[i]
+		if credential.SSHPublicKey == nil {
+			continue
+		}
+
+		if credential.SSHPublicKey.Source.Secret == nil || credential.SSHPublicKey.Source.Secret.SecretName == "" {
+			continue
+		}
+		secretName := credential.SSHPublicKey.Source.Secret.SecretName
+
+		if credential.SSHPublicKey.PropagationMethod.QemuGuestAgent == nil {
+			continue
+		}
+
+		users := credential.SSHPublicKey.PropagationMethod.QemuGuestAgent.Users
+		for j := range users {
+			if users[j] == user {
+				result = append(result, secretName)
+				break
+			}
+		}
+	}
+	return result
+}
+
+func secretContainsKey(secretData map[string][]byte, key string) bool {
+	for _, data := range secretData {
+		lines := strings.Split(string(data), "\n")
```

Comment:
Is it correct checking the ssh key line by line? I mean we could have new lines or not but the key is the same? Shouldn't we trim the data and eliminate then `\n` and then compare?

Reply:
One `data` entry in the Secret contains a file, that can contain multiple keys, each on it's own line.
That's why this loop checks each line separately.